src/lib/functions/netlify-function.js

```diff
@@ -0,0 +1,113 @@
+const BACKGROUND_SUFFIX = '-background'
+
+class NetlifyFunction {
+  constructor({
+    config,
+    errorExit,
+    functionsDirectory,
+    mainFile,
+    name,
+    projectRoot,
+    runtime,
+    timeoutBackground,
+    timeoutSynchronous,
+  }) {
+    this.config = config
+    this.errorExit = errorExit
+    this.functionsDirectory = functionsDirectory
+    this.mainFile = mainFile
+    this.name = name
+    this.projectRoot = projectRoot
+    this.runtime = runtime
+    this.timeoutBackground = timeoutBackground
+    this.timeoutSynchronous = timeoutSynchronous
+
+    // Determines whether this is a background function based on the function
+    // name.
+    this.isBackground = name.endsWith(BACKGROUND_SUFFIX)
+
+    // List of the function's source files. This starts out as an empty set
+    // and will get populated on every build.
+    this.srcFiles = new Set()
+  }
+
+  // The `build` method transforms source files into invocable functions. Its
+  // return value is an object with:
+  //
+  // - `srcFilesDiff`: Files that were added and removed since the last time
+  //    the function was built.
+  async build({ cache }) {
+    const buildFunction = await this.runtime.getBuildFunction({
+      cache,
+      config: this.config,
+      errorExit: this.errorExit,
+      func: this,
+      functionsDirectory: this.functionsDirectory,
+      projectRoot: this.projectRoot,
+    })
+
+    this.buildQueue = typeof buildFunction === 'function' ? buildFunction(this) : undefined
```

Comment:
Good catch! Initially I was planning on making the build function optional, because the JS runtime doesn't really build anything when a function builder is not being used. However, I ended up creating an "artificial" build function for this case (https://github.com/netlify/cli/pull/2826/files#diff-d18e598ea7298f3e008d51148f7839ea5199a5d378d79de6e2272622663a57a7R40-R46), so this check is no longer necessary.

I've removed it in https://github.com/netlify/cli/pull/2826/commits/b34edeeee532b8df46fcc1ee4cac267c1612770b.